waspc/src/Generator/Job.hs

```diff
@@ -0,0 +1,29 @@
+module Generator.Job
+    ( Job
+    , JobMessage (..)
+    , JobMessageData (..)
+    , JobOutputType (..)
+    , JobType (..)
+    ) where
+
+import           Control.Concurrent (Chan)
+import           System.Exit        (ExitCode)
+
+
+-- | Job is an IO action that communicates progress by writing messages to given channel
+--   until it is done, when it returns exit code.
+type Job = Chan JobMessage -> IO ExitCode
+
+data JobMessage = JobMessage
+    { _data    :: JobMessageData
+    , _jobType :: JobType
+    }
+    deriving (Show)
+
+data JobMessageData = JobOutput String JobOutputType
+                    | JobExit ExitCode
+    deriving (Show)
+
+data JobOutputType = Stdout | Stderr deriving (Show)
+
+data JobType = WebApp | Server deriving (Show)
```

Comment:
hm interesting. this could also be encoded at the type lvl right? :D but ok let's keep it simple, I am just wondering what is the standard way to do stuff like this.

Reply:
Hm I don't know actually. Channel needs to receive both types of messages, so I don't think it could.